api/envoy/service/tap/v2alpha/common.proto

```diff
@@ -0,0 +1,88 @@
+option java_multiple_files = true;
+
+// [#protodoc-title: Common tap configuration]
+
+// Tap configuration.
+message TapConfig {
+  // The match conditions. One or more match configurations can be specified. If any of them
+  // match the data source being tapped, a tap will occur, with the result written to the
+  // configured output.
+  repeated MatchConfig match_configs = 1 [(validate.rules).repeated .min_items = 1];
+
+  // The tap output configuration. If a match configuration matches a data source being tapped,
+  // a tap will occur and the data will be written to the configured output.
+  OutputConfig output_config = 2 [(validate.rules).message.required = true];
+
+  // [#comment:TODO(mattklein123): Rate limiting]
+}
+
+// Tap match configuration. Within a single match configuration, all conditions are evaluated as
+// a logical AND. Logical OR matching can be implemented via multiple match configurations in the
+// top level tap configuration.
+message MatchConfig {
+  // Match ID used for correlating tap data with the match that produced it.
+  string match_id = 1 [(validate.rules).string.min_bytes = 1];
+
+  // HTTP match configuration. Only useful for HTTP aware tap extensions such as the
+  // :ref:`HTTP tap filter <config_http_filters_tap>`.
+  HttpMatchConfig http_match_config = 2;
+}
+
+// HTTP match configuration.
+message HttpMatchConfig {
+  // HTTP request match configuration.
+  HttpRequestMatchConfig request_match_config = 1;
+
+  // HTTP response match configuration.
+  HttpResponseMatchConfig response_match_config = 2;
+}
+
+// HTTP request match configuration.
+message HttpRequestMatchConfig {
+  // HTTP request headers to match.
+  repeated api.v2.route.HeaderMatcher headers = 1;
+}
+
+// HTTP response match configuration.
+message HttpResponseMatchConfig {
+  // HTTP response headers to match.
+  repeated api.v2.route.HeaderMatcher headers = 1;
+}
+
+// Tap output configuration.
+message OutputConfig {
+  // Output sinks for tap data. Currently a single sink is allowed in the list. Once multiple
+  // sink types are supported this constraint will be relaxed.
+  repeated OutputSink sinks = 1 [(validate.rules).repeated = {min_items: 1, max_items: 1}];
+
+  // [#comment:TODO(mattklein123): Output filtering. E.g., certain headers, truncated body, etc.]
+}
+
+// Tap output sink configuration.
+message OutputSink {
+  oneof output_sink_type {
+    option (validate.required) = true;
+
+    // Tap output will be streamed out the :http:post:`/tap` admin endpoint.
+    //
+    // .. attention::
+    //
+    //   It is only allowed to specify the streaming admin output sink if the tap is being
+    //   configured from the :http:post:`/tap` admin endpoint. Thus, if an extension has
+    //   been configured to receive tap configuration from some other source (e.g., static
+    //   file, XDS, etc.) configuring the streaming admin output type will fail.
+    StreamingAdminSink streaming_admin = 1;
```

Comment:
While I now grok the idea of potentially multiple sinks and a per-sink config, I wonder if we want this one to be more like `ResponseStreamSink`; this would apply in the gRPC case if the response comes back on bidi as well as the admin case.

Reply:
Let me think about this. I think that could work. I guess the question is whether we would do a bidi for this. My thinking on how to handle gRPC is actually two different APIs: 1) an xDS for fetching tap configs, and 2) and access log / metrics service like API for pushing taps. This would allow these things to be decoupled. I.e., xDS could be used to configure a file sink, or a static config could be used to configure a gRPC push sink, for example.

Reply:
I can see that model working, but I also think that a common use case will be a client connecting to Envoy via gRPC. E.g. imagine I write a Wireshark-like client, I want to listen to some Envoy. I point my client at this Envoy and want it to stream stuff back to me. In the past whenever I've wanted to TAP, this has been the debug flow, rather than configuring a server endpoint to receive the TAP on. Network asymmetries could make that impractical.

I do get the long term direction you're heading here and the want to support a push as well as pull gRPC model, since the TAP might be used as part of a more sophisticated monitoring solution that is not under human debugger control, just pointing out that both models probably would be good to handle for gRPC as well.

Reply:
Right that makes sense, but for the case you described, isn't this still over admin? Meaning, instead of a `ResponseStreamSink` wouldn't we actually allow the admin /tap endpoint to initiate a gRPC bidi stream, and then in the sink config we would have an output type which could be JSON/gRPC/etc.?

Reply:
Here's an example, based on a true story. You have the debug Wireshark-client under the control of a user in network X, Envoy is running in network Y. There are NAT/firewall between X/Y which allows X to initiate connections into Y but not vice versa. You need to be able to initiate all connections from the client in X.

Reply:
We are on the same page. My point is that I think your use case is still an admin request/response since it's a direct connection into Envoy. I'm suggesting we would implement as options to the admin sink that describe whether it's JSON, gRPC, etc. Given that, does it make sense to change the name or leave it as `StreamingAdminSink` ?

Reply:
Yeah, we can leave it as is I guess. I actually think the gRPC connection would not be to the `/tap` endpoint, it would be to some proto specified service implied endpoint, but that's a more general "what should the admin gRPC interface look like?".